Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean

```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
 
 end Limits
 
+end CategoryTheory
+
+namespace CategoryTheory
+
+universe v v₂ u u₂
+
+namespace Limits
+
+/-- The type of objects for the diagram indexing reflexive (co)equalizers -/
+inductive WalkingReflexivePair : Type where
+  | zero
+  | one
+  deriving DecidableEq, Inhabited
+
+open WalkingReflexivePair
+
+/-- The type of morphisms for the diagram indexing reflexive (co)equalizers -/
+inductive WalkingReflexivePairHom : (WalkingReflexivePair → WalkingReflexivePair → Type)
+  | left : WalkingReflexivePairHom one zero
+  | right : WalkingReflexivePairHom one zero
+  | reflexion : WalkingReflexivePairHom zero one
+  | left_reflexion : WalkingReflexivePairHom one one
+  | right_reflexion : WalkingReflexivePairHom one one
+  | id (X : WalkingReflexivePair) : WalkingReflexivePairHom X X
+  deriving DecidableEq
+
+attribute [-simp, nolint simpNF] WalkingReflexivePairHom.id.sizeOf_spec
+
+open WalkingReflexivePairHom
+
+/-- Composition of morphisms in the diagram indexing reflexive (co)equalizers -/
+def WalkingReflexivePairHom.comp :
+    ∀ { X Y Z : WalkingReflexivePair } (_ : WalkingReflexivePairHom X Y)
+      (_ : WalkingReflexivePairHom Y Z), WalkingReflexivePairHom X Z
+  | _, _, _, id _, h => h
+  | _, _, _, h, id _ => h
+  | _, _, _, reflexion, left => id zero
+  | _, _, _, reflexion, right => id zero
+  | _, _, _, reflexion, right_reflexion => reflexion
+  | _, _, _, reflexion, left_reflexion => reflexion
+  | _, _, _, left, reflexion => left_reflexion
+  | _, _, _, right, reflexion => right_reflexion
+  | _, _, _, right_reflexion, right_reflexion => right_reflexion
+  | _, _, _, right_reflexion, left_reflexion => right_reflexion
+  | _, _, _, right_reflexion, right => right
+  | _, _, _, right_reflexion, left => right
+  | _, _, _, left_reflexion, left => left
+  | _, _, _, left_reflexion, right => left
+  | _, _, _, left_reflexion, right_reflexion => left_reflexion
+  | _, _, _, left_reflexion, left_reflexion => left_reflexion
+
+theorem WalkingReflexivePairHom.id_comp
+    {X Y : WalkingReflexivePair} (g : WalkingReflexivePairHom X Y) : comp (id X) g = g := by
+  cases g <;> rfl
+
+theorem WalkingReflexivePairHom.comp_id
+    {X Y : WalkingReflexivePair} (f : WalkingReflexivePairHom X Y) : comp f (id Y) = f := by
+  cases f <;> rfl
+
+theorem WalkingReflexivePairHom.assoc {X Y Z W : WalkingReflexivePair}
+    (f : WalkingReflexivePairHom X Y) (g: WalkingReflexivePairHom Y Z)
+    (h : WalkingReflexivePairHom Z W) : comp (comp f g) h = comp f (comp g h) := by
+  cases f <;> cases g <;> cases h <;> rfl
+
+instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
+  Hom := WalkingReflexivePairHom
+  id := id
+  comp := comp
+  comp_id := comp_id
+  id_comp := id_comp
+  assoc := assoc
```

Comment:
As these lemmas should not be needed outside of the definition of the category structure, it would suggest doing this:
```suggestion
instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
  Hom := WalkingReflexivePairHom
  id := id
  comp := comp
  comp_id f := by cases f <;> rfl
  id_comp f := by cases f <;> rfl
  assoc f g h := by cases f <;> cases g <;> cases h <;> rfl
```